go/consensus/tendermint/apps/governance/governance.go

```diff
@@ -149,12 +149,36 @@ func (app *governanceApplication) BeginBlock(ctx *api.Context, request types.Req
 		"upgrade", ud,
 	)
 
-	// Check if we are running the correct version for the upgrade, if not - trigger a consensus halt.
-	if err := ud.EnsureCompatible(); err != nil {
-		ctx.Logger().Error("not running updated binary",
-			"err", err,
-		)
-		return upgrade.ErrStopForUpgrade
+	// Check if the upgrade descriptor is installed in the node and has been executed.
+	if upgrader := ctx.AppState().Upgrader(); upgrader != nil {
+		switch pu, err := upgrader.GetUpgrade(ctx, ud); err {
+		case nil:
+			// Upgrade exists, make sure it is in the process of being applied.
+			if !pu.HasStage(upgrade.UpgradeStageStartup) {
+				ctx.Logger().Error("upgrade pending but not being applied",
+					"handler", ud.Handler,
+					"epoch", ud.Epoch,
+					"upgrade_height", pu.UpgradeHeight,
+					"last_completed_stage", pu.LastCompletedStage,
+				)
+				return upgrade.ErrStopForUpgrade
+			}
+		case upgrade.ErrUpgradeNotFound:
```

Comment:
This prevents locally canceling a consensus upgrade (2/3+ nodes locally canceling an upcoming pending upgrade), but i think this is fine as in case an already accepted upgrade needs to be reverted/aborted manual intervention is needed anyway so this can be achieved otherwise (e.g. stopping the nodes at the upgrade epoch and doing a "dump-restore" without the upgrade)

Reply:
Yeah, especially since there is also an option to cancel an upgrade via governance (should we consider reducing `upgrade_cancel_min_epoch_diff`?).

Reply:
Yeah that would make sense